api/api.go

```diff
@@ -0,0 +1,93 @@
+package api
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"time"
+
+	"github.com/hashicorp/consul-terraform-sync/event"
+)
+
+const (
+	defaultAPIVersion = "v1"
+
+	statusHealthy      = "healthy"
+	statusDegraded     = "degraded"
+	statusCritical     = "critical"
+	statusUndetermined = "undetermined"
+)
+
+// API supports api requests to the cts biniary
+type API struct {
+	store   *event.Store
+	port    int
+	version string
+	srv     *http.Server
+}
+
+// NewAPI create a new API object
+func NewAPI(store *event.Store, port int) *API {
+	mux := http.NewServeMux()
+
+	// retrieve task status for a task-name
+	mux.Handle(fmt.Sprintf("/%s/%s/", defaultAPIVersion, taskStatusPath),
+		newTaskStatusHandler(store, defaultAPIVersion))
+	// retrieve all task statuses
+	mux.Handle(fmt.Sprintf("/%s/%s", defaultAPIVersion, taskStatusPath),
+		newTaskStatusHandler(store, defaultAPIVersion))
+
+	srv := &http.Server{
+		Addr:         fmt.Sprintf(":%d", port),
+		WriteTimeout: time.Second * 15,
+		ReadTimeout:  time.Second * 15,
+		IdleTimeout:  time.Second * 60,
+		Handler:      mux,
+	}
+
+	return &API{
+		port:    port,
+		store:   store,
+		version: defaultAPIVersion,
+		srv:     srv,
+	}
+}
+
+// Serve starts up and handles shutdown for the http server to serve
+// API requests
+func (api *API) Serve(ctx context.Context) {
+	go func() {
+		if err := api.srv.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+			// go ahead and stop cts from continuing
+			log.Fatalf("error starting api server at '%d': '%s'\n", api.port, err)
+		}
+	}()
+
+	log.Printf("[INFO] (api) server started at port %d", api.port)
+
+	ctxShutDown, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
```

Comment:
With the only thing under here a goroutine, wouldn't this defer be in a race with the goroutine's use of it?

Reply:
The deferred cancel that is.

Reply:
I see what you mean about the two cancels racing. I'm new to working directly with go routines / context 🤔 I thought it would make the most sense to remove the cancel in the go routine. Let me know what you think!

Reply:
When thinking about where to put it you first think of where/how you're using the context paired with it. You want the cancel to pair up with the end of the context's life. With it where you currently have the cancel defer, it will cancel the context to soon, possibly even before it gets used, as it will cancel when it hits the end of the function and will not wait for that goroutine to get started (even if it did, it'd still cancel it way before that time out is reachec).

I think you want the `cancel()` only in the goroutine. Either in  a defer, or just before the return call. That will cancel the context as it's use ends. In fact you could move creating the context into the goroutine, just before the for loop... I think that is probably what I'd do. Then you can just create and cancel in a defer together as you have it, which is the more common pattern.

Reply:
Thanks for your guidance here and taking the time to explain this! That makes sense. I updated to create the context and cancel within the go-routine (see last commit).

Reply:
This looks good. Made one new related comment below.